robota_core/repository.py

```diff
@@ -17,43 +18,62 @@
 from robota_core.github_tools import GithubServer
 from robota_core.string_processing import string_to_datetime
 
+GITLAB_DIFF_FILES_PER_PAGE = 20
 
 class Branch:
     """An abstract object representing a git branch.
 
-    :ivar id: Name of branch.
+    :ivar name: Name of branch.
     :ivar id: Commit id that branch points to.
+    :ivar project_url: base URL of the project.
     """
-    def __init__(self, branch, source: str):
+    def __init__(self, branch, source: str, project_url: str = None):
```

Comment:
Yes, but the project URL is already embedded in the web_url attribute of the specific objects we are retrieving, and is already in the right format.  So the work you are currently doing on line 51 of your version of `repository.py` is not needed, because the full URL (including the project URL) is already made for you.  E.g. see the JSON from:

https://gitlab.com/api/v4/projects/60747754/repository/branches/main

which is:

```
{
  "name": "main",
  "commit": {
    "id": "61ba593d2ce0884dfe7516dd4253b3d51a0543dd",
    "short_id": "61ba593d",
    "created_at": "2024-08-10T23:23:13.000+01:00",
    "parent_ids": [
      "273d16fd7bd370af2cb74609b3f899d50e3235b7"
    ],
    "title": "Provide more informative README",
    "message": "Provide more informative README\n",
    "author_name": "XXX",
    "author_email": "XXX",
    "authored_date": "2024-08-10T23:23:13.000+01:00",
    "committer_name": "XXX",
    "committer_email": "XXX",
    "committed_date": "2024-08-10T23:23:13.000+01:00",
    "trailers": {

    },
    "extended_trailers": {

    },
    "web_url": "https://gitlab.com/robota-suite/test-public-repo/-/commit/61ba593d2ce0884dfe7516dd4253b3d51a0543dd"
  },
  "merged": false,
  "protected": true,
  "developers_can_push": false,
  "developers_can_merge": false,
  "can_push": true,
  "default": true,
  "web_url": "https://gitlab.com/robota-suite/test-public-repo/-/tree/main" <-- HERE
}
```

This last component of the JSON (the `web_url` attribute) contains the URL ready made, in the same form as the link your code is constructing in line 51 (shown just after my comment).

I'm suggesting we use this URL, and then we don't need anyone to pass the project's URL to this bit of the code.

Reply:
Apologies, I understand now. This works fine for Gitlab, but Github does not have a `web_url` equivalent on the 'list all branches' endpoint. `_links.html` is only returned when querying a _single branch_.

Reply:
Are we using the "list all branches" endpoint here?  If so, we can just issue another call to get the information for the specific named branch.  That should have the information needed, I think?